### ISSUE TITLE
[FEATURE REQUEST] PrusaLink HTTPS Support

### ISSUE THREAD

**Issue by Redstonecrafter0** [prusa3d/Prusa-Firmware-Buddy#4215]
### Printer model

MK4S

### Describe the feature

HTTPS would make PrusaLinks authentication secure as the credentials would be encrypted.
Secure namespaces like .dev would work to access to 3D printer. I use a .dev domain locally. At least Google Chrome does not allow http access on secure namespaces.

In an IoT vlan without internet access Prusa Connect would be no alternative.

### Expected functionality

The certchain and private key could be imported from the USB flash drive. Alternatively they cloud stay stored on the flash drive and be loaded on boot since when HTTPS support is desired one would upload via HTTPS anyway.
If there is no user provided (trusted) certificate there should be a untrusted one generated.

**Comment by Redstonecrafter0** [comment]
bump

**Comment by moritzmhmk** [comment]
bump

**Comment by Redstonecrafter0** [comment]
Hi, I appreciate your time and effort regarding this issue.

However I have trouble understanding the following:
- Prusa Connect is already using TLS 1.2 between the printer and the cloud. [Prusa Blog](https://blog.prusa3d.com/prusa-connect-a-network-solution-for-remote-control-of-3d-printers_90364/)
- TLS 1.2 for an embedded HTTPS server would be challenging because of power and memory constraints.

In my eyes this is the same problem to solve where one of them is already solved. 

If I am missing something, I'd really appreciate your insights. Thanks again!

**Comment by Redstonecrafter0** [comment]
I took the liberty of testing the server using wrk. With more than two concurrent connections the requests/sec reduces. It's true that browsers open several connections at the same time for faster loading but in a local network it does not matter since TCP SYNs are retried and the latency is negligible.

The option for security should not be sacrificed for speed. It should at least be tested in a live environment if an HTTPS server would be viable. Otherwise I only see a self-hosted instance of Prusa Connect as an option which I highly doubt will happen.

**Comment by Redstonecrafter0** [comment]
bump

**Comment by Redstonecrafter0** [comment]
bump

**Comment by Redstonecrafter0** [comment]
bump

**Comment by Redstonecrafter0** [comment]
Can I at least convince you to implement this because of the Cyber Resilience Act?

**Comment by amoose136** [comment]
HTTPS is kinda sucky on a local network unless that network also has access to the internet so chain of trust can work and you can use something like [Let's Encrypt](https://en.wikipedia.org/wiki/Let%27s_Encrypt). Otherwise, you're in the land of self-signed certs which every major browser will rightfully warn you about when you try and connect to... IE it's mostly security theater if you don't have chain of trust to a Certificate Authority. That said, if you ever want to do something like use a websocket to communicate to a browser, https is the only way to go. This would be advantageous to enable certain things like low-overhead Server Sent Events. [A standard ESP32 can support being an HTTPS server](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/protocols/esp_https_server.html) so there _should be_ no hardware limitation here.

**Comment by Redstonecrafter0** [comment]
There is no need for a security theater but i wanna use a trusted cert with a chain of trust to my own CA. I'm not against an ACME client but initially I think any way to secure the communication to the 3D printer by using https would be a major win.

**Comment by tinus-github** [comment]
You can actually install and run a reverse proxy and see that it works fine, except for very limited circumstances where you get links to the unproxied instance. But the performance is not prohibitively bad.

**Comment by amoose136** [comment]
The fact remains there isn’t yet a good solution in the case that the network is air gapped (which Prusa printers support with caveats). That’s the reason that even new routers don’t have their admin/configuration pages served over https.

**Comment by Redstonecrafter0** [comment]
Can't confirm. For me the Prusa printer is the only device with a webinterface incapable of serving https. My router and even my paper printer are capable of https. Also looking forward the CRA effectively mandates https by default for routers, printers, you name it.

**Comment by tinus-github** [comment]
I don’t think we need to discuss these arguments. Yes, situations exist in which it is hard or considered unnecessary to use https. That doesn’t change the fact that situations also exist in which it is preferable to use https. This feature request is about these situations.

Imagine if it was possible to do use https! You could just choose for yourself to use it, or not use it.

**Comment by fphammerle** [comment]
not stale

**Comment by Redstonecrafter0** [comment]
bump

**Comment by moritzmhmk** [comment]
Prusa Connect Local https://www.prusa3d.com/p/prusa-connect-local might make this issue obsolete - at least for my use case 🤔
